File: pset6/credit/credit.py
from cs50 import get_int
from math import floor

# Card number input
card_num = get_int("Number: ")

# Parameter
count = 0
digits = 0
adding = 0
checksum = 0
tmp = card_num
count_num = str(card_num)

# Digit count from input
count = int(len(count_num))

#Luhn's algorithm DEBUGGING IS WAITING
for i in range(1, count + 1):
    digits = floor(tmp % 10)

    if i % 2 == 0:
        digits *= 2

        if digits > 9:
            digits -= 9
        adding = digits
    else:
        adding = digits

    checksum += floor(digits)
    tmp /= 10

# Create company's ID
temp_id = card_num
while temp_id > 100:
    temp_id /= 10

company_id = floor(temp_id)

# Checking: AMEX, MASTERCARD, VISA
if checksum % 10 == 0 and count == 13 and company_id in [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]:
    print("VISA")

elif checksum % 10 == 0 and count == 15 and company_id in [34, 37]:
    print("AMEX")

elif checksum % 10 == 0 and count == 16:

    if company_id in [51, 52, 53, 54, 55]:
        print("MASTERCARD")

    if company_id in [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]:
        print("VISA")

else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
//card_num defines as user's input
    long card_num = get_long("Number: ");
//-----------------------------------------------------------//
//Parameters define
    int digit = 0;
    int count = 0;
    int sum = 0;
    long temp_num = card_num;
    long count_num = card_num;
//-----------------------------------------------------------//
//Finding digits of the input
    while (count_num > 0)
    {
        count++;
        digit = count_num / 10;
        count_num /= 10;
    }
//-----------------------------------------------------------//
//Luhn's algorithm
    for (int i = 1; i < count + 1; i++)
    {
        int digit_num = temp_num % 10;
        if (i % 2 == 0)
        {
            digit_num *= 2;
            if (digit_num > 9)
            {
                digit_num -= 9;
            }
        }
        sum += digit_num;
        temp_num /= 10;
    }
//-----------------------------------------------------------//
//Create the company_id
    long temp_id = card_num;
    while (temp_id > 100)
    {
        temp_id /= 10;
    }
    int company_id = temp_id;
//-----------------------------------------------------------//
//Checking card type: AMEX, MASTERCARD, VISA
    if ((sum % 10 == 0) && (count == 13) && (company_id >= 40 && company_id <= 49))
    {
        printf("VISA\n");
    }
    else if ((sum % 10 == 0) && (count == 15) && (company_id == 34 || company_id == 37))
    {
        printf("AMEX\n");
    }
    else if ((sum % 10 == 0) && (count == 16))
    {
        if (company_id >= 51 && company_id <= 55)
        {
        printf("MASTERCARD\n");
        }
        else if (company_id >= 40 && company_id <= 49)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}
